### PR TITLE
Add Multistory and Preview Decorator Support

### PR DESCRIPTION
**Pull request by usulpro**
Issue: #1335

## What I did

Add simple API for switching storyKinds to the Multistory Mode by adding `:` to storyKind

example:

```js
storiesOf('Buttons:', module)
  .add('normal small', ...)
  .add('normal big', ...)
  .add('normal enormous', ...)
```

will show all stories on one preview page. All stories still are selectable in the Stories Panel.
You can use decorators to customize how each story will look.
It works on any hierarchy level as well:

```js
storiesOf('Buttons:.normal', module)
  .add('normal small', ...)
  .add('normal big', ...)
  .add('normal enormous', ...);

storiesOf('Buttons:.accent', module)
  .add('accent small', ...)
  .add('accent big', ...)
  .add('accent enormous', ...);
```

will create a common `Buttons` page with both `normal` and `accent` buttons 

## How to test

Github Page: https://github.com/UsulPro/storybook-multistory

run `cra-kitchen-sink` example

### CODE REVIEW

**Comment by ndelangen** [comment]
Awesome work @UsulPro !

**Comment by UsulPro** [comment]
image

**Comment by UsulPro** [comment]
I added two options:

```js
setOptions({
  multistorySeparator: /:/,
  previewDecorator: tileDecorator,
});
```

The first one sets the reg.exp. to determine multistories

The second allows to pass custom React Component to the root of preview:

```js
const tileDecorator = stories =>
  <div>
    <h1>Button Tiles</h1>
    <div
      style={{
        display: 'flex',
        flexWrap: 'wrap',
        alignItems: 'stretch',
      }}
    >
      {stories}
    </div>
  </div>;
```

how to test:

Run kitchen-sink. Check out `Buttons Guide -> tile:` section

**Comment by UsulPro** [comment]
I guess it's ready!

@ndelangen @shilman @igor-dv could you review?

File: app/react/package.json [hunk 1]
```diff
@@ -19,12 +19,14 @@
...
     "dev": "DEV_BUILD=1 nodemon --watch ./src --exec 'npm run prepublish'",
-    "prepublish": "node ../../scripts/prepublish.js"
+    "prepublish": "node ../../scripts/prepublish.js",
+    "test": "../../node_modules/.bin/jest --watch"
   },
   "dependencies": {
     "@storybook/addon-actions": "^3.1.6",
     "@storybook/addon-links": "^3.1.6",
     "@storybook/addons": "^3.1.6",
+    "@storybook/addon-options": "^3.1.6",
```
Comment (igor-dv) [inline app/react/package.json]: Maybe I'm missing the point, so fix me if I'm wrong. 
You are adding here a dependency for the `addon-options` -> hence a coupling to it. AFAIK `addon-actions` and `addon-links` are already deprecated as a default addons of the app. So it makes me unsure regarding this design.. (After that you will need to add it to `app/react-native` and `app/vue` )

Comment (UsulPro) [inline app/react/package.json]: @igor-dv, it's a very good question!

addon-options is a very special package - it doesn't provide any additional functionality like `actions` and `lilnks`. We use it to get access to UI API. 

So maybe better to think that it's not an addon at all!? :thinking: 

For Storybook users it's just one kind of API. Common use:
```js
import { configure, storiesOf } from '@storybook/react';
import { setOptions } from '@storybook/addon-options'; // why I need one more package?
```

Nevertheless, the only my reason here is to provide such API:
```js
setOptions({
  hierarchySeparator: /\/|:\//, // regular expression to separate stories nesting
  multistorySeparator: /:/, // regular expression to separate multistories section
  previewDecorator: stories => <div>{stories}</div>, // root decorator for preview
});
```

I guess it's much more comfortable for consumers to have all similar APIs in one place, than for example:

```js
import { setOptions } from '@storybook/addon-options';
import { setPreferences } from '@storybook/react';

setOptions({
  hierarchySeparator: /\/|:\//, // regular expression to separate stories nesting
});

setPreferences({
  multistorySeparator: /:/, // regular expression to separate multistories section
  previewDecorator: stories => <div>{stories}</div>, // root decorator for preview
});

```

The last example is more consistent with the project design, but it can :suspect: >> :boom: >> :feelsgood:

File: addons/options/src/preview/index.js [hunk 2]
```diff
@@ -7,6 +7,19 @@ export function init() {
   // NOTE nothing to do here
 }
 
+let previewOptions = () => {};
+
+function regExpStringify(exp) {
```
Comment (igor-dv) [inline addons/options/src/preview/index.js]: cool idea =)

File: app/react/src/client/preview/render.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 /* global document */
+/* eslint react/prop-types: 0 */
```
Comment (igor-dv) [inline app/react/src/client/preview/render.js]: why ?

File: app/react/src/client/preview/render.js [hunk 4]
```diff
@@ -65,37 +175,63 @@ export function renderMain(data, storyStore) {
...
-  if (element.type === undefined) {
-    const error = {
-      title: `Expecting a valid React element from the story: "${selectedStory}" of "${selectedKind}".`,
-      description: stripIndents`
-        Seems like you are not returning a correct React element from the story.
-        Could you double check that?
-      `,
-    };
-    return renderError(error);
+  // const element = multiStoriesSeparator
```
Comment (igor-dv) [inline app/react/src/client/preview/render.js]: redundant comment ?

**Review by igor-dv — COMMENTED** [review]
Maybe it worth to open a discussion regarding the addon-option. If it's not an addon it should be somewhere else =)

**Comment by ndelangen** [comment]
I love this feature, but the API, I'm not a fan of. I have reservations about using the first argument (string).

I do not want to end up supporting tons of options, settings, arguments, presets, loaders, methods inside a string.

I propose (discussed with @shilman) an generic API for setting options on a story:
```js
storiesOf('a/b/c', module).setOptions({ previewMultiple: true })
```

**Review by ndelangen — CHANGES_REQUESTED** [review]
I'd like the API to change

**Comment by shilman** [comment]
To add to @ndelangen 's comment, `setOptions({ ... })` globally can behave like `addDecorator` to set options in the global scope, and `setOptions` on a story kind can add those options to the kind or the stories within the kind.

Still discussing this with @ndelangen @igor-dv @tmeasday @alexandrebodin and no consensus except that everybody is concerned about over-using the string argument.

**Comment by design1online** [comment]
Sweet!! You made my day :-)

File: app/react/package.json (hunk 1 above)
Comment (shilman) [inline app/react/package.json]: FWIW i agree with @igor-dv that `addon-options` is a strange package and we should consider moving the functionality into `addons`. I've opened an issue https://github.com/storybooks/storybook/issues/1430

I believe this is complementary to my proposal lower down to create a `storiesOf(...).setOptions({ ... })` method for configuring showing multiple preview windows.

**Comment by UsulPro** [comment]
@ndelangen @shilman Thank you for your feedback!

I love the idea of generic API for setting options on a story. And I believe as we implement it we start to receive tons of feature requests for options to support inside this API. So string argument will be saved from bloating! :)

Ok, I will start to change multistory API to `setOptions` style. Maybe better to submit a separate PR with `setOptions` feature first.

I see there will be some issues/questions related to using it for multistory so  let's consider it as well:

1.a) "string" API for multistories allows to union several "storykinds" into one multistory group. Especially it looks cool when it goes with hierarchy feature - so you can use the left panel to navigate through this multistories page.

Example:

```js
storiesOf('Buttons:.simple', module)
  .add('normal small', ...)
  .add('normal big', ...)
  .add('normal enormous', ...);

storiesOf('Buttons:.emoji', module)
  .add('accent small', ...)
  .add('accent big', ...)
  .add('accent enormous', ...);
```

[demo](https://usulpro.github.io/storybook-multistory/?selectedKind=Buttons%20Guide:.emoji&selectedStory=normal%20small&full=0&down=1&left=1&panelRight=1&downPanel=storybook%2Fnotes%2Fpanel ) - notice how it joins "simple" and "emoji" sections. 

And you can still use different decorators to set the own style for each section.

A possible solution for this could be such way of using setOptions :

```
storiesOf(...)
  .setOptions({ previewMultiple: 'Buttons'})
```

1.b) Dividing *storyKind* by a separator we can pass then this data to decorators inside `context` argument and use it for better output.

same solution as above (`storiesOf(...).setOptions({ previewMultiple: 'Buttons'})`) allows to divide storyKing into two parts

2) Via RegExp we can easily turn all storyKinds to multistory this way:

setOptions({
  multistorySeparator: /$/,
});

or create one big multistory from the whole storybook

setOptions({
  multistorySeparator: /^/,
});

but I guess we can have the same effect if allow to pass RegExp to our new API:

```
storiesOf(...)
  .setOptions({ previewMultiple: /Buttons/})
```

it's going to run even better if we take @shilman's idea of local/global `setOptions`! So we could just set it globally:

```
  .setOptions({ previewMultiple: /$/}) // switch all storyKinds to multistory
```

What do you think?

**Comment by UsulPro** [comment]
It worth to mention that one more thing I added in this PR is three additional arguments are passed into a function invoked by a decorator. 

**Previous API:**

```js
stories = stories.addDecorator(function decoratorFn(storyfn,
  {  kind, story },
) { /* Return a decorator */ })
```

**New API:**

```js
stories = stories.addDecorator(function decoratorFn(storyfn,
  { kind, story, kindRoot, selectedStory, onStoryDidMount }, // <---
) { /* Return a decorator */ })
```

additional props are:

`kindRoot` - the common part of a `storyKind`s in case when we union some storyKinds into one multistory (if not in multistory mode it's same as kind)

`selectedStory` - the story selected at the Stories Panel (if not in multistory mode it's same as story) 

`onStoryDidMount(function storyDidMount(elmentID))` - a function that takes a callback function storyDidMount as an agrument, which will be invoked after all stories will be mounted. storyDidMount will be invoked with an argument `elmentID` [string] - an ID attribute of <div> node containing the current story

### Why we need this?

<details>
  <summary> To customize appearance and behavior  </summary>


These options allow to customize appearance and behavior of multistories exclusively via decorators on the stories side:

- by default multistories have such structure:

```html
<body>
  <div id="root">
    <PreviewDecorator>
      <div id="story1_name">
         {storyFn1()}
      </div>
      <div id="story2_name">
         {storyFn2()}
      </div>
      { ... }
    </PreviewDecorator>
  </div>
</body>
```

and default previewDecorator is:
```js
previewDecorator: stories =>
<div className="stories-root">
  {stories}
</div>
```

- if you `.addDecorator` to storiesOf you'll have:

```html
<body>
  <div id="root">
    <PreviewDecorator>
      <div id="story1_name">
         {decoratorA(storyFn1, context)}
      </div>
      <div id="story2_name">
         {decoratorA(storyFn2, context)}
      </div>
      { ... }
    </PreviewDecorator>
  </div>
</body>
```

- and if you union several storykinds into one multistory like `storiesOf(A)` and `storiesOf(B)`:

```html
<body>
  <div id="root">
    <PreviewDecorator>
      <div id="story1a_name">
         {decoratorA(storyFn1a, context)}
      </div>
      <div id="story2a_name">
         {decoratorA(storyFn2a, context)}
      </div>
      <div id="story1b_name">
         {decoratorB(storyFn1b, context)}
      </div>
      <div id="story2b_name">
         {decoratorB(storyFn2b, context)}
      </div>
      { ... }
    </PreviewDecorator>
  </div>
</body>
```

- you can use `context` for example to:

  - display title of a story

  - display a section (like `simple` / `emoji` in example)

  - highlight selected story

  - scroll to selected story:

```js
context.onStoryDidMount(id => {
  const currentStoryHolder = document.getElementById(id);
  if (currentStoryHolder) currentStoryHolder.scrollIntoView();
});
```

- you might want to use another container for your multistories.

Example:

```js
const flexDecorator = stories =>
  <div  style={{ display: 'flex' }}>
      {stories}
  </div>


setOptions({
  previewDecorator: flexDecorator,
});
```

</details>

File: app/react/src/client/preview/render.js (hunk 4 above)
Comment (UsulPro) [inline app/react/src/client/preview/render.js]: sure!

File: addons/options/src/preview/index.js (hunk 2 above)
Comment (UsulPro) [inline addons/options/src/preview/index.js]: maybe it worth to submit a separate PR with this feature to have it before we finish with this PR? @igor-dv is it still actual?

Comment (igor-dv) [inline addons/options/src/preview/index.js]: oh, missed this comment.. I think it's relevant..

**Comment by ndelangen** [comment]
@UsulPro Want to pair and finish this?

**Comment by ndelangen** [comment]
I Merged in release/3.3 and fixed it up a bit.

Some code is disabled.. but it seems to work.
I'm still feeling the `:` syntax is not that great. But I'm also ready to put some effort into this to get it ready for 3.3 release.

@shilman ?

**Comment by InkeyesLT** [comment]
Hey, very nice work on this feature, can we expect it any time soon?

**Comment by UsulPro** [comment]
@InkeyesLT Thanks! Actually, I was a bit busy for some time, but I hope I can come back to it. 

@Norbert if your offer is still relevant, it would be great if we make this happen together :) I described different possible API in my previous comment and [here](https://gist.github.com/UsulPro/14c84d734f45ca9692ff7517b5d4f0a3). Anyway, I'm open to discuss and continue this anytime when you're ready. Maybe next week?

**Comment by henningmu** [comment]
Hi @ndelangen and @UsulPro could you please tell me what's up with this PR / feature? Will it never be part of storybook or is development continued elsewhere?

**Comment by mariepw** [comment]
@UsulPro @henningmu @ndelangen FYI I was looking for this feature, as "workaround" I just created an overview page for all my stories : 

```
import * as React from "react";
import { storiesOf, getStorybook } from "@storybook/react";

const allStories = getStorybook();

const AllStories = () => (
  <>
    {
      allStories.map((stories) => {
        return <> {
          stories.stories.map((story) => {
            return <>{ story.render() }</>
          })
        } </>
      })
    }
  </>
);

storiesOf("Overview", module).add("all", () => <AllStories />); 
```

Using Storybook `5.0.0`
